core/ipc/twin/common/src/main/java/org/opennms/core/ipc/twin/common/AbstractTwinSubscriber.java

```diff
@@ -230,72 +136,173 @@ protected TwinUpdate mapTwinResponseToProto(byte[] responseBytes) {
     }
 
-    public MinionIdentity getMinionIdentity() {
-        return minionIdentity;
+    public Identity getIdentity() {
+        return this.identity;
     }
 
-    private String getLocation() {
-        if (minionIdentity != null) {
-            return minionIdentity.getLocation();
+    private static class Value {
+        public final String sessionId;
+        public final int version;
+
+        public final JsonNode value;
+
+        private Value(final String sessionId,
+                      final int version,
+                      final JsonNode value) {
+            this.sessionId = Objects.requireNonNull(sessionId);
+            this.version = version;
+            this.value = Objects.requireNonNull(value);
         }
-        return null;
     }
 
-    private class SessionImpl<T> implements Closeable {
-
+    private class Subscription {
         private final String key;
-        private final Consumer<T> consumer;
-        private final Class<T> clazz;
 
-        public SessionImpl(String key, Class<T> clazz, Consumer<T> consumer) {
-            this.key = key;
-            this.clazz = clazz;
-            this.consumer = consumer;
-        }
+        private final Set<Consumer<JsonNode>> consumers = Sets.newConcurrentHashSet();
+
+        private Value value = null;
 
-        @Override
-        public void close() throws IOException {
-            sessionMap.remove(key, this);
-            LOG.info("Closed session with key {} ", key);
+        /**
+         * A tasked scheduled to retry an outstanding request.
+         *
+         * This is a one-shot task re-executing the request. The task will be scheduled when a request is emitted and
+         * will be canceled when an update is accepted. Setting this to {@code null} shows that there is no outstanding
+         * request.
+         */
+        private ScheduledFuture<?> retry = null;
+
+        private Subscription(final String key) {
+            this.key = Objects.requireNonNull(key);
         }
 
-        public void accept(byte[] objValue) throws IOException {
-            final T value = objectMapper.readValue(objValue, clazz);
-            LOG.trace("Updated consumer with key {}", key);
-            consumer.accept(value);
+        /**
+         * Consume the subscription.
+         *
+         * Adds a consumer to the subscription. Incoming values will be forwarded to the passed consumer until the
+         * returned value is closed. If there is a value already available for the subscription, the consumer will be
+         * called with this value immediately.
+         *
+         * @param clazz The class of the value to consume
+         * @param consumer The consumer accepting the values
+         * @param <T> The class of the value to consume
+         * @return a Closable, stopping the consumption when closed
+         */
+        public synchronized <T> Closeable consume(final Class<T> clazz, final Consumer<T> consumer) {
```

Comment:
In case that several consumers register at around the same time and no value is available yet, multiple requests are sent. Maybe the subscriber could use information that a request is already underway. A non-null `retry` future would be an indicator for that.

Reply:
Moved to retry-cancel logic to `accept` to prevent loop if nothing changed and added proposed check.